test/e2e/autoscaler.go

```diff
@@ -239,67 +298,127 @@ func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput)
+			})
+
+			By("Creating workload that forces the system to scale up")
+			framework.AddScaleUpDeploymentAndWait(ctx, framework.AddScaleUpDeploymentAndWaitInput{
+				ClusterProxy: workloadClusterProxy,
+			}, input.E2EConfig.GetIntervals(specName, "wait-autoscaler")...)
+
+			By("Checking the MachinePool is scaled up")
+			mpScaledUpReplicas := mpOriginalReplicas + 1
+			framework.AssertMachinePoolReplicas(ctx, framework.AssertMachinePoolReplicasInput{
+				Getter:             input.BootstrapClusterProxy.GetClient(),
+				MachinePool:        clusterResources.MachinePools[0],
+				Replicas:           mpScaledUpReplicas,
+				WaitForMachinePool: input.E2EConfig.GetIntervals(specName, "wait-autoscaler"),
+			})
+
+			By("Disabling the autoscaler")
+			framework.DisableAutoscalerForMachinePoolTopologyAndWait(ctx, framework.DisableAutoscalerForMachinePoolTopologyAndWaitInput{
+				ClusterProxy:                  input.BootstrapClusterProxy,
+				Cluster:                       clusterResources.Cluster,
+				WaitForAnnotationsToBeDropped: input.E2EConfig.GetIntervals(specName, "wait-controllers"),
+			})
+
+			By("Checking we can manually scale up the MachinePool")
+			// Scale up the MachinePool. Since autoscaler is disabled we should be able to do this.
+			mpExcessReplicas := mpScaledUpReplicas + 1
+			framework.ScaleMachinePoolTopologyAndWait(ctx, framework.ScaleMachinePoolTopologyAndWaitInput{
+				ClusterProxy:        input.BootstrapClusterProxy,
+				Cluster:             clusterResources.Cluster,
+				Replicas:            mpExcessReplicas,
+				WaitForMachinePools: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+				Getter:              input.BootstrapClusterProxy.GetClient(),
+			})
+
+			By("Checking enabling autoscaler will scale down the MachinePool to correct size")
+			// Enable autoscaler on the MachinePool.
+			framework.EnableAutoscalerForMachinePoolTopologyAndWait(ctx, framework.EnableAutoscalerForMachinePoolTopologyAndWaitInput{
+				ClusterProxy:                input.BootstrapClusterProxy,
+				Cluster:                     clusterResources.Cluster,
+				NodeGroupMinSize:            mpNodeGroupMinSize,
+				NodeGroupMaxSize:            mpNodeGroupMaxSize,
+				WaitForAnnotationsToBeAdded: input.E2EConfig.GetIntervals(specName, "wait-autoscaler"),
+			})
+
+			By("Checking the MachinePool is scaled down")
+			// Since we scaled up the MachinePool manually and the workload has not changed auto scaler
+			// should detect that there are unneeded nodes and scale down the MachinePool.
+			framework.AssertMachinePoolReplicas(ctx, framework.AssertMachinePoolReplicasInput{
+				Getter:             input.BootstrapClusterProxy.GetClient(),
+				MachinePool:        clusterResources.MachinePools[0],
+				Replicas:           mpScaledUpReplicas,
+				WaitForMachinePool: input.E2EConfig.GetIntervals(specName, "wait-controllers"),
+			})
+
+			if input.ScaleToAndFromZero {
```

Comment:
Is this feature specified for MachinePools? (they don't have an InfraMachineTemplate)

Mabye let's focus on MD for now and add a disclaime to the ScaleToAndFromZero input field?

Reply:
👍

Reply:
hmm according to the doc, it should work if the annotations are set.

Also for machinepools, if the inframachinepool provides `.status.capacity` (e.g. DockerMachinePool) it should work too.

Reply:
https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210310-opt-in-autoscaling-from-zero.md#machineset-and-machinedeployment-annotations

Due to that I removed it